### PR TITLE
Disqualify built-in trait impl if it seems likely to overlap in an unsound way with a blanket impl

### PR DESCRIPTION
**Pull request by compiler-errors**
Hi T-types. 

**TL;DR:** This PR fixes an unsoundness having to do with `dyn Trait` trait object built-in impls overlapping with user-written blanket impls by disqualifying the built-in impl when the trait has associated types and it fails a simple heuristic (based off of where clauses[^1]) to determine whether or not the built-in impl overlaps with the user-written impl.
[^1]: More about this below.

## What is the unsoundness

We currently do not detect when an object's built-in impl overlaps with a blanket impl for an unsized self type, like `impl<T: ?Sized> Foo for T`. The trait solver prefers object candidates over impl candidates because of traits like `Any`, which require dispatching to their virtual function to operate corerctly.

This preference can be side-stepped when we're in a generic context, and when the blanket impl differs from the built-in impl in an associated type, this can be used to unsoundly implement a transmutation function:

```rust
trait Object<U> {
    type Output;
}
impl<T: ?Sized, U> Object<U> for T {
    type Output = U;
}
fn foo<T: ?Sized, U>(x: <T as Object<U>>::Output) -> U {
    x
}

fn transmute<T, U>(x: T) -> U {
    foo::<dyn Object<U, Output = T>, U>(x)
}
```

In this case, the user-written `Object<U, Output = U>` impl overlaps with the built-in `Object<U, Output = T>` from the dyn.

## How does this PR fix this?

We implement a heuristic query called `trait_has_impl_which_may_shadow_dyn` that takes a *trait def id* and the *principal def id* of a dyn type. It only considers traits with associated types, since those are the only traits that are inherently unsound with overlapping impls.

For the given trait def id, we check all of its impls:
  * We use deep-reject to unify the self type. If it fails, then we disqualify the impl.
  * If it has a `Sized` predicate that shares the same self type as the impl, then it is disqualified.
  * For each *other* trait predicate that shares the self type, we recurse, checking if that trait may apply to the dyn.
      * If we find no impl, then the impl is disqualified.
  * If the impl remains, then we consider the impl to be shadowed and return true.

If this heuristic returns true, then it will disqualify the built-in dyn impl so we always prefer the user-written impl. This check must be *complete*, i.e. it may never return false if an impl may apply, since we will then fail to disqualify an overlapping impl which gives us a way to side-step this check. However, it doesn't need to be *sound* i.e. it may return false-positives and may consider more impls to overlap than necessary. This will result in trait errors, but this does not happen in practice according to the crater run results.

## Regressions:

[*transmutter-0.1.0*](https://crater-reports.s3.amazonaws.com/pr-132289-1/try%234ce0d30f2c24854a6a87cad4608888f9c18a94db/reg/transmutter-0.1.0/log.txt) - This is just an example of the unsoundness in question.
[*hooks-yew-1.0.0-alpha.2*](https://crater-reports.s3.amazonaws.com/pr-132289-1/try%234ce0d30f2c24854a6a87cad4608888f9c18a94db/reg/hooks-yew-1.0.0-alpha.2/log.txt) - This is a legitimate regression. lcnr minimized it down to:

```rust
trait Trait<T> {
    fn method(&self);
}
trait Erased<T> {
    type Assoc;
    fn erased(&self);
}
impl<T> Trait<T> for dyn Erased<T, Assoc = ()> {
    fn method(&self) {
        self.erased();
    }
}
impl<T, H: Trait<T> + ?Sized> Erased<T> for H {
    type Assoc = ();
    fn erased(&self) {
        self.method();
    }
}

fn break_me(hr: Box<dyn for<'a> Erased<&'a (), Assoc = ()>>) {
    hr.erased();
}
```

Which fails because our dyn overlap heuristic is not smart enough, since it doesn't consider higher-ranked placeholder errors when unifying the impl signature...

### Always prefer the blanket impl

This would technically be the most sound option, since we could just always shadow the built-in object impl by a blanket if it applies.

However, this is likely unworkable, because although we can use some sort of intrinsic or built-in compiler hack to fix the behavior of `Any`, I don't believe that we could apply this hack to user-built analogues of `Any` or other traits that generally prefer to dispatch to the built-in object impl over the blanket, and so this would a breaking change in general. 

This PR compromises by only shadowing the built-in impl when it has associated types, as it would be unsound in these cases if the associated type differed between the impls.

### Always consider the impl as being "specialized" by the built-in object impl

This means that any blanket impl that overlaps with a built-in object impl is considered to be "further specializable", and it would mean that we cannot project an associated type in an example like:

```rust
trait Mirror {
    type Assoc;
}

impl<T: ?Sized> Mirror for T {
    type Assoc;
}

fn foo<T: ?Sized>(x: &T) -> &<T as Mirror>::Assoc { x }
```

since that `<T as Mirror>::Assoc` type may return a different associated type for the "specializing" built-in dyn impl.

### CODE REVIEW

**Comment by compiler-errors** [comment]
@craterbot check

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 1]
```diff
@@ -478,3 +479,58 @@ fn report_conflicting_impls<'tcx>(
...
+        {
+            return;
+        }
+
+        if let ty::Alias(ty::Projection, alias_ty) = self_ty.kind()
+            && tcx
+                .item_super_predicates(alias_ty.def_id)
+                .iter_identity()
+                .filter_map(|clause| clause.as_trait_clause())
+                .any(|bound| bound.def_id() == sized_trait)
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:527]: we may want to elaborate here, already done implicitly for the `caller_bounds` above

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/specialize/mod.rs:527]: `item_super_predicates` is elaborated

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/specialize/mod.rs:527]: explicit_item_super_predicates <- the non-elaborated one

**Comment by lcnr** [comment]
crater analysis https://hackmd.io/burOEs9MQdelPgk4frE76g

**Comment by compiler-errors** [comment]
for later:

`craterbot check p=1 crates=https://crater-reports.s3.amazonaws.com/pr-132289/retry-regressed-list.txt`

**Comment by compiler-errors** [comment]
@craterbot check p=1 crates=https://crater-reports.s3.amazonaws.com/pr-132289/retry-regressed-list.txt

File: compiler/rustc_trait_selection/src/traits/project.rs [hunk 2]
```diff
@@ -817,6 +817,12 @@ fn assemble_candidates_from_object_ty<'cx, 'tcx>(
     let env_predicates = data
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/project.rs:817]:
```suggestion
    let object_ty_predicates = data
```

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 3]
```diff
@@ -595,3 +598,128 @@ fn report_conflicting_impls<'tcx>(
...
+    )
+}
+
+fn trait_has_impl_inner<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    target_trait_def_id: DefId,
+    target_self_ty: SimplifiedType<DefId>,
+    elaborated_supertraits: &FxHashSet<DefId>,
+    seen_traits: &mut FxHashSet<DefId>,
+    for_nested_impl: bool,
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:637]: pls change to an `enum ForNestedImpl`

or wait: alternative:

if we duplicate the "is `Sized` check" in `fn trait_has_impl_which_may_shadow_dyn` and have it call an extracted function `fn check_user_written_impls`, which then recurse into `trait_has_impl_inner`, we don't need `ForNestedImpl` at all.

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 4]
```diff
@@ -595,3 +598,128 @@ fn report_conflicting_impls<'tcx>(
...
+    if !tcx
+        .associated_items(target_trait_def_id)
+        .in_definition_order()
+        .any(|item| item.kind == ty::AssocKind::Type)
+    {
+        return false;
+    }
+
+    let target_self_ty =
+        principal_def_id.map_or(SimplifiedType::MarkerTraitObject, SimplifiedType::Trait);
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:616]:
```suggestion
        SimplifiedType::Trait(principal_def_id.unwrap());
```
do we get here with auto-traits? I don't think so, as we'd always return `true`  in this case

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 5]
```diff
@@ -595,3 +598,128 @@ fn report_conflicting_impls<'tcx>(
...
+            return;
+        }
+
+        let self_ty = tcx
+            .impl_trait_ref(impl_def_id)
+            .expect("impl must have trait ref")
+            .instantiate_identity()
+            .self_ty();
+
+        if simplify_type(tcx, self_ty, TreatParams::InstantiateWithInfer)
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:679]:
```suggestion
        // We only care about impls which are implemented for the trait object
        // itself and blanket impls which may apply.
        if simplify_type(tcx, self_ty, TreatParams::InstantiateWithInfer)
```

**Review by lcnr — COMMENTED** [review]
While reviewing this I've written down the desired behavior of `trait_has_impl_which_may_shadow_dyn` myself to make sure I don't just accept the impl as is. This may be helpful as a slightly different perspective on what we're doing here:

We only disable the builtin impls for traits with associated types, as only these may have unsound overlap. We correctly handle overlapping methods by always using the builtin impl for them, necessary for `<dyn Any as Any>::type_id`.

For such traits

- there exists an impl which does not require `Sized` on a non-rigid self-type
- unless this impl has a where-bound on the self-type which requires the self-type to implement a trait which is known to not hold:

the trait may hold if it is not `Sized` and
- it is an auto-trait, or
- the trait is a builtin trait implemented for `dyn Trait (+ Auto)*`, or
- the trait is part of the elaborated predicates list, i.e. a super-trait, or
- there exists an user-written impl which applies to this trait object
    - either implemented directly for `dyn Trait (+ Auto)*`
    - blanket impl
        - unless this impl has a where-bound on the self-type which requires the self-type to implement a trait which is known to not hold -> recurse
        - or the self-type is an alias whose item bounds cannot apply -> recurse

**Review by lcnr — COMMENTED** [review]
ah, addition concerns + required tests

- ~~**REQUIRED CHANGE**: I think we need to extend the check to disable `dyn Trait: Trait` even if that passes this check and only `dyn Trait: Super` is disabled, otherwise the implication can happen inside of a nested function. also needs test~~ *edit: already happens implicitly as we have to check all super predicates of trait objects as part of their builtin impl anyways, pls add test*
```rust
trait Super {
    type Assoc;
}

impl<T: ?Sized> Super for T {
    type Assoc = usize;
}

trait Trait: Super<Assoc = &'static str> {}

fn via_impl<T: ?Sized>() -> <T as Super>::Assoc {
    0
}

fn via_trait<T: Trait + ?Sized>() -> &'static str {
    via_impl::<T>()
}

fn main() {
    println!("{}", via_trait::<dyn Trait>());
}
``` 
- add UI test
  - case where we disable the builtin impl `Erased` from the FCP
  - the impl does apply but the self-type is a projection whose item bound do not apply to the trait object
  - we disable the builtin `Super` impl for `dyn Trait`
  - disable both `Trait`  and `Super` for `dyn Trait: Super` and there is `impl<T: Trait> Super for T`? we have to disable both `Trait` and `Super` here, don't we?

**Comment by lcnr** [comment]
after this (and another crater run? :thinking:) I'd start the FCP

**Review by jackh726 — COMMENTED** [review]
>we may also want to emit a warn by default lint for all impls which disable the builtin object impl 🤔 even if that lint will mostly be unactionable 😅 potentially even regardless of whether we actually disable the impl because the trait has an associated type as adding such an impl will mean that adding an associated type will cause your trait to not be object safe anymore which is likely surprising

I'm really curious of the fallout of this is (i.e. how widespread is this pattern). There are effectively four levels of strictness here:
1) Error on all impls that may overlap with the builtin impl, regardless of it they have an associated type
2) Error on all impls that may overlap with the builtin impl and have an associated type, whether or not that associated type can overlap the builtin associated type
3) Error on all impls that may overlap with the builtin impl and have an associated type that *can* overlap with the builtin associated type
4) Error on all impls that may overlap with the builtin impl and have an associated type that *does* overlap with the builtin associated type.

I'm not sure if there's a meaningful difference between 3 and 4, but there may be in what we can *prove*.

I imagine 1 and 2 are probably non-starters in terms of breakage, but 3 or 4 may be worth doing at the coherence level. This is instead of the PR as-is, which *allows* overlapping impls but *disables* the builtin impl.

Of course, 3 and 4 get harder to implement when you consider GATs (which aren't yet dyn compatible but will be), but maybe not unapproachably harder.

I think this PR is compatible with later switching to one of the 4 sets of rules above, since those are I think more strict than the rules in this PR. So, I think it's probably worth landing this, since is better in terms of soundness.

However, I do think that my comments regarding user type annotations being effectively ignored should be addressed. It isn't critical for *correctness*, so *maybe* not worth blocking this PR on, but imo does provide a bad user experience so it's something I would *really* like to see before the PR lands.

File: tests/ui/coherence/overlap-dyn-impl-in-codegen.rs [hunk 6]
```diff
@@ -11,6 +11,7 @@ impl<A: ?Sized> MyTrait for A {
 
 fn main() {
     bug_run::<dyn MyTrait<Target = u8>>();
+    //~^ ERROR the size for values of type
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply tests/ui/coherence/overlap-dyn-impl-in-codegen.rs:14]: > Now that we disqualify the builtin impl, this implication should just always be true, right?

Correct.

> erroring on line 21 feels right

In another Rust, maybe, but I don't believe it's feasible to fix this now. Making blanket impls not apply generically causes tons of breakage.

**Comment by steffahn** [comment]
This PR introduces a soundness regression:
```rust
type A = &'static [usize; 1];
type B = &'static [usize; 100];

type DynSomething = dyn Something<Assoc = A>;

trait Super {
    type Assoc;
}
impl Super for Foo {
    type Assoc = A;
}

trait IsDynSomething {}
impl IsDynSomething for DynSomething {}

impl<T: ?Sized> Super for T
where
    T: IsDynSomething,
{
    type Assoc = B;
}

trait Something: Super {
    fn method(&self) -> Self::Assoc;
}

struct Foo;
impl Something for Foo {
    fn method(&self) -> Self::Assoc {
        &[1337]
    }
}

fn main() {
    let x = &Foo;
    let y: &DynSomething = x;

    let short_arr: A = x.method();
    let long_arr: B = y.method();
    println!("{:?}", short_arr);
    println!("{:?}", long_arr);
}
```
fails on stable/nightly, UB on this branch

**Comment by compiler-errors** [comment]
of fucking course it does lmao

**Comment by compiler-errors** [comment]
Pretty sure this is just @lcnr's:

> disable both Trait and Super for dyn Trait: Super and there is impl<T: Trait> Super for T? we have to disable both Trait and Super here, don't we?

**Comment by steffahn** [comment]
I don't know, I stopped trying to understand the description and discussion here, and instead just downloaded the branch and played with it until it broke 😇

**Comment by compiler-errors** [comment]
@steffahn: Ya im just talking to myself (and lcnr maybe)

**Comment by steffahn** [comment]
Here's another problematic case I could find. It's a 2-crates setup:

First crate (e.g. your `lib.rs`):
```rust
pub type A = &'static [usize; 1];
pub type B = &'static [usize; 100];

pub trait Trait<P> {
    type Assoc;
}

pub type Dyn<P> = dyn Trait<P, Assoc = A>;

pub trait AvoidOverlap<It: ?Sized> {}
impl<P> AvoidOverlap<Dyn<P>> for () {}

pub trait LocallyUnimplemented<P> {}

impl<P, T: ?Sized> Trait<P> for T
where
    T: LocallyUnimplemented<P>,
    (): AvoidOverlap<T>
{
    type Assoc = B;
}

pub fn function<P>() -> <Dyn<P> as Trait<P>>::Assoc {
    &[1337]
}
```
Second crate (e.g. your `main.rs`):
```rust
use nam_of_lib::{function, Dyn, LocallyUnimplemented};

struct Param;

impl LocallyUnimplemented<Param> for Dyn<Param> {}

fn main() {
    let arr: &[usize; 100] = function::<Param>();
    println!("{:?}", arr);
}
```
example output on this branch:
```rust
[1337, 9223371974577750009, 9223361443317940208, 7598807758576447331, 8387236760591036015, 72340172838076673, 72340172838076673, 8028075845441778529, 2338606692308099182, 8295751895439843369, 6926649950808009068, 7935476084319219301, 7236850772546188399, 8385534785046799406, 8031998611105869945, 8245656750015082344, 7308324465818165605, 7235433442201987451, 9041083404643824416, 18445953872741353664, 18445908449167217680, 2334106421097295465, 7308604897068083558, 7954799936718922098, 2339731488442490980, 506381209866536711, 506381209866536711, 1, 2, 1, 0, 2314885530818453536, 2338600897117954080, 651061555542690057, 651061555542690057, 7881131626781303854, 8031998611039609697, 7598543892949132624, 7234304299306872431, 7309993588911404115, 8245935277856024934, 578721382704613384, 578721382704613384, 7021797314962419566, 7305790156120810855, 0, 0, 4716517092481066049, 7308324465885667702, 2332992598738301472, 2322278944636232038, 361700864190383365, 361700864190383365, 3978425819141910832, 7378413942531504440, 49087, 0, 9223372039002259456, 9223372039002259456, 1, 1, 479142256574474, 479520213808129, 479545983611853, 499977143039031, 502300720351447, 502300720351447, 502300720351447, 500269200819878, 500531193825495, 499951373240094, 502300720350900, 502300720351447, 502300720351054, 502300720350990, 504538398312238, 504693017136191, 506316514774122, 504538398313184, 506045931834080, 504538398313220, 504538398313184, 504538398313184, 504538398313184, 504538398313184, 504538398313184, 504538398313184, 504538398313184, 504538398313184, 504538398313184, 504538398313184, 504693017136191, 504693017135876, 504693017135876, 506230615427844, 504693017135840, 504693017135876, 504538398313220, 511964396767968, 510783280763022]
```
error on stable/nightly
```rust
error[E0308]: mismatched types
 --> src/main.rs:8:30
  |
8 |     let arr: &[usize; 100] = function::<Param>();
  |              -------------   ^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 100 elements, found one with 1 element
  |              |
  |              expected due to this
```

**Comment by steffahn** [comment]
Here's another interesting case, with (up to) 3 crates:

## code

(uses the standard library as the first crate:)
```rust
pub trait Index<Idx: ?Sized> {
    type Output: ?Sized;
    fn index(&self, index: Idx) -> &Self::Output; // <- the method is actually irrelevant for this issue
}
```
(any object-safe trait with a type parameter and associated type works)

second crate (e.g. your `lib.rs`):
```rust
use std::ops::Index;

pub trait Trait {
    fn f(&self)
    where
        dyn Index<(), Output = ()>: Index<()>;
    //  rustc (correctly) determines ^^^^^^^^ this bound to be true
}

pub fn call(x: &dyn Trait) {
    x.f(); // so we can call `f`
}
```

third crate (e.g. your `main.rs`):
```rust
use std::ops::Index;

use play2_stage::{call, Trait};

trait SubIndex<I>: Index<I> {}

struct Param;

trait Project {
    type Ty: ?Sized;
}
impl Project for () {
    type Ty = dyn SubIndex<(), Output = ()>;
}

impl Index<Param> for <() as Project>::Ty {
    type Output = ();

    fn index(&self, _: Param) -> &() {
        &()
    }
}

struct Struct;

impl Trait for Struct {
    fn f(&self)
    where
        for<'a> dyn Index<(), Output = ()>: Index<()>,
        // now rustc believes that this bound ^^^^^^^ is false
    {
        println!("hello!");
    }
}

fn main() {
    call(&Struct); // <- hence, the method `f` isn't part of the vtable

    // ^^^ segfaults (calls null function pointer from vtable)
}
```

## remark
I'm saying "up to 3" because above, the first two crates *can* be the same.

The 3-crate layout does however better clarify the separation of concerns. In particular, one could say: “the `Index` trait isn't doing anything suspicious”.

On stable `rustc`, this code runs fine and prints `hello`.

**Comment by compiler-errors** [comment]
Yeah, I dont want to work on this anymore. It's a terrible approach and it's not making any progress in FCP anyways. A more feasible strategy at the very least needs to more accurately model the logic of "is this trait implemented via an impl" but likely should use the preexisting machinery we have today.

**Comment by jackh726** [comment]
@steffahn can you open a PR to add those examples as tests?

**Comment by steffahn** [comment]
yes, I will do that
